### ISSUE TITLE
ChunkLoadError: Loading chunk XX failed (during add-on installation)

### ISSUE THREAD

**Issue by willdurand** [mozilla/addons-frontend#8259]
There are a lot of errors like below since the last push.

https://sentry.prod.mozaws.net/operations/addons-frontend-amo-prod/issues/6015921/

```
ChunkLoadError: Loading chunk 49 failed.
(error: https://addons-amo.cdn.mozilla.net/amo-i18n-pl-5895ff0e3199834fe71a.js)
  at location (./src/locale/fr/amo.js?0bc5:7:1)
  at __webpack_require__ (./src/amo/components/LanguagePicker/index.js:36:28)
  at e/< (./src/amo/components/LanguagePicker/index.js:36:13)
  at G (./src/amo/utils/errors.js:14:14)
  at a (./src/amo/components/LanguagePicker/index.js:57:14)
...
(16 additional frame(s) were not displayed)
```

**Comment by kumar303** [comment]
Sentry suggests that these errors may be preventing add-on installation. Because of this and that 6.8K users are affected thus far, I'm marking it a P1.

**Comment by willdurand** [comment]
> Sentry suggests that these errors may be preventing add-on installation. Because of this and that 6.8K users are affected thus far, I'm marking it a P1.

I tried clicking the install button but I cannot reproduce...

**Comment by willdurand** [comment]
We believe that the `react-transition-group` lib might be the problem. We merged an update in #8226 and we found references of a `SwitchTransition` component in the Sentry logs. This lib has been updated in #8241, **after** the tag. The update mentions the `SwitchTransition` component.

We think that upgrading the lib might fix this problem.

**Comment by kumar303** [comment]
> We believe that the `react-transition-group` lib might be the problem. 

Specifically, the stack traces begin with this:

<img width="735" alt="Screenshot 2019-07-05 14 21 55" src="https://user-images.githubusercontent.com/55398/60742667-d956be80-9f33-11e9-85ef-533af8416144.png">

There might be additional clues on what was fixed in 4.2.1 here: https://github.com/reactjs/react-transition-group/pull/516

If upgrading to 4.2.1 doesn't stop the errors, we will try downgrading https://github.com/mozilla/addons-frontend/pull/8226

**Comment by kumar303** [comment]
tag prepped for a hotfix: https://github.com/mozilla/addons-frontend/compare/2019.07.04...2019.07.04-1

**Comment by willdurand** [comment]
The hotfix has been deployed in production but we're still seeing the same Sentry events 😞

**Comment by willdurand** [comment]
> If upgrading to 4.2.1 doesn't stop the errors, we will try downgrading #8226

I am going to try that in `2019.07.04-2`, see: 2019.07.04-2

**Comment by ioanarusiczki** [comment]
@willdurand Had some errors at first on a new profile - FF67 - Win10 on stage which I can no longer reproduce now. 

![errors](https://user-images.githubusercontent.com/33448286/60805148-f683cb00-a187-11e9-9c70-a2810eb16118.png)

**Comment by willdurand** [comment]
> I am going to try that in `2019.07.04-2`, see: [2019.07.04-1...2019.07.04-2](https://github.com/mozilla/addons-frontend/compare/2019.07.04-1...2019.07.04-2)

We're still seeing the same Sentry events after the second hotfix has been deployed. I believe it has nothing to do with `react-transition-group`...

It could be caused by https://github.com/mozilla/addons-frontend/commit/5c1cfa7505b02cd77124326476852848af05d114, maybe.

**Comment by willdurand** [comment]
Maybe related threads:

- https://github.com/webpack/webpack/issues/5630
- https://github.com/webpack/webpack/issues/7502#issuecomment-436619373

**Comment by willdurand** [comment]
Sentry stats:

![Screen Shot 2019-07-08 at 15 37 47](https://user-images.githubusercontent.com/217628/60814722-6ef18880-a196-11e9-8a8a-fb590086a3ff.png)
![Screen Shot 2019-07-08 at 15 38 07](https://user-images.githubusercontent.com/217628/60814723-6ef18880-a196-11e9-913b-ca30075b3312.png)

- Device: 61% are smartphone
- OS: 40% Windows 10, 24% Windows 7

Still, I cannot reproduce with a fresh Windows 10 + FF 67.

**Comment by AlexandraMoga** [comment]
I've also tested on an Android 8.0 with Fx67 and I could not reproduce the issue

**Comment by willdurand** [comment]
Most (all?) stack traces have a reference to the lang picker. It's weird because the lang picker does not use CSS transitions and updating the value of the lang picker reloads the page.

**Comment by willdurand** [comment]
Also, re: polyfills, the polyfills we're using in addons-frontend are likely not needed with FF 67+, so I am not sure that the updated polyfills are the cause.

Some babel deps have been updated too during last week's tag, so a behavior may have changed there too.

**Comment by kumar303** [comment]
This seems to affect a high number of users but there are no actual reports from users, which is atypical considering our userbase. A likely explanation of this is that it's not breaking the site in any visible way. Thus, I'm dropping priority. 

It remains a mystery. None of us can reproduce it. The webpack issues up above suggest that it could be an intermittent problem due to the user's network or it could be caused by a browser extension on the client machine which is blocking webpack from loading resources. 

It would be nice to figure this out simply because something changed in this deployment to cause it.

**Comment by muffinresearch** [comment]
I wonder if sentry could be logging something on unload e.g. after lang picker interaction and during the start of the reload?

**Comment by willdurand** [comment]
I see that this error appeared after last week's push: https://sentry.prod.mozaws.net/operations/addons-frontend-amo-prod/issues/6015991/. It seems very similar to the error linked to this GitHub issue. There is a reference to `core-js`, which has been updated in the last push according to the migration guide: https://babeljs.io/docs/en/babel-polyfill, https://babeljs.io/blog/2019/03/19/7.4.0#core-js-3-7646-https-githubcom-babel-babel-pull-7646 and https://github.com/zloirock/core-js#babelpolyfill.

That being said, I only noticed now that the babel polyfill requires `core-js@2`, so the best explanation I have right now is that there is a subtle BC break between `core-js@2` and `core-js@3`.

Another issue to take into account: I believe the `core-js` update was necessary to be able to upgrade other dependencies. For example, Storybook uses its own babel/webpack setup and it requires `core-js@3` (if I am not mistaken).

**Comment by willdurand** [comment]
`NS_ERROR_FAILURE` seems to be a general FF error according to https://developer.mozilla.org/en-US/docs/Mozilla/Errors:

> NS_ERROR_FAILURE (0x80004005)
>    This is the most general of all the errors and occurs for all errors for which a more specific error code does not apply.

**Comment by kaushalinfosys** [comment]
I am using AsyncComponent HOC to lazy load chunks, and was facing same issue.
the work around I did is, identifying the error and make a hard reload once.
```
.catch(error => {
        if (error.toString().indexOf('ChunkLoadError') > -1) {
          console.log('[ChunkLoadError] Reloading due to error');
          window.location.reload(true);
        }
      });
```

the full HOC file looks like this,
```
export default class Async extends React.Component {
  componentWillMount = () => {
    this.cancelUpdate = false;
    this.props.load
      .then(c => {
        this.C = c;
        if (!this.cancelUpdate) {
          this.forceUpdate();
        }
      })
      .catch(error => {
        if (error.toString().indexOf('ChunkLoadError') > -1) {
          console.log('[ChunkLoadError] Reloading due to error');
          window.location.reload(true);
        }
      });
  };

  componentWillUnmount = () => {
    this.cancelUpdate = true;
  };

  render = () => {
    const props = this.props;
    return this.C ? (
      this.C.default ? (
        <this.C.default {...props} />
      ) : (
        <this.C {...props} />
      )
    ) : null;
  };
}
```

**Comment by willdurand** [comment]
Interesting, thanks @kaushalinfosys. Would you know the root cause, though?

**Comment by kaushalinfosys** [comment]
> Interesting, thanks @kaushalinfosys. Would you know the root cause, though?

Yes, the root cause is webpack changes the chunk file name on every build considering the code has been changed, see the setting in `webpack.config.prod.js` file,
`chunkFilename: 'static/js/[name].[chunkhash:8].chunk.js',`

If we keep our application tab open and meanwhile some PROD deployment is done, and if we try to navigate the page without refreshing the page, this error will occur because the chunk references in main.[hash].js is old, but the lazy load file name is already changed on server. So main file can not find that chunk.

**Comment by willdurand** [comment]
mm, thanks. That's interesting because I did not think about that. I am confused because we have errors coming almost all the time in Sentry and we deploy only once a week 🤔

**Comment by hannadrehman** [comment]
@kaushalinfosys i initially thought the same , but i have observed that even if there is no deployment i get this chunkLoadError quite sometime in a day.

**Comment by bobsilverberg** [comment]
@willdurand Do you know where things stand with these errors? Are we still seeing a lot of them, or did the problem magically go away?

**Comment by willdurand** [comment]
> @willdurand Do you know where things stand with these errors? Are we still seeing a lot of them, or did the problem magically go away?

still there AFAIK
